src/compiler/checker.ts

```diff
@@ -32200,7 +32200,10 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
         // accessor, or optional method.
         const assignmentKind = getAssignmentTargetKind(node);
         if (assignmentKind === AssignmentKind.Definite) {
-            return removeMissingType(propType, !!(prop && prop.flags & SymbolFlags.Optional));
+            // since this is used as an assignment target we always want to remove the missing type
+            // usually this is done conditionally based on the `prop.flags`
+            // the prop might not be available here though if the assignment is done to a dynamic property
+            return removeMissingType(propType, /*isOptional*/ true);

```

Comment:
This makes `isOptional` parameter somewhat misleading here.

The idea is that if we have `missingType` in the `propType` here we should just always remove it. It could only be obtained here from that access expression itself (I think?) so if we write to that access expression it should be always safe to just try to remove it - without even checking if the prop was optional or not. 

To fix the off-putting usage of this parameter I could:
- rename the parameter
- leave it as is since there is an inline comment that mentions why it's done this way
- inline `removeMissingType` here